src/detect-parse.c

```diff
@@ -1495,6 +1495,15 @@ int DetectSignatureSetAppProto(Signature *s, AppProto alproto)
         return -1;
     }
 
+    /* since AppProtoEquals is quite permissive wrt dcerpc and smb, make sure
+     * we refuse `alert dcerpc ... smb.share; content...` explicitly. */
+    if (alproto == ALPROTO_SMB && s->alproto == ALPROTO_DCERPC) {
```

Comment:
This will reject `alert smb (dcerpc.keyword; smb.keyword)` even if you wanted to keep it, no ?

`alert smb` sets `ALPROTO_SMB`

dcerpc.keyword then sets `ALPROTO_DCERPC`

Then smb.keyword fails

Reply:
in this case `alproto` should be `ALPROTO_DCERPC` and `s->alproto` should be `ALPROTO_SMB`

Reply:
example
```json
{
  "raw": "alert smb any any -> any any (msg:\"test\"; dcerpc.iface:367abb81-9844-35f1-ad32-98f038001003; dcerpc.opnum:12; dcerpc.stub_data; content:\"|15 00 00 00 00 00 00 00 15 00 00 00|\"; within:32; pcre:\"/^(?:[A-Z]\\x00){20}\\x00\\x00/R\"; content:\"|15 00 00 00 00 00 00 00 15 00 00 00|\"; distance:6; within:12; pcre:\"/^(?:[A-Z]\\x00){20}\\x00\\x00/R\"; content:\"|03 00 00 00|\"; distance:10; within:4; sid:2;)",
  "id": 2,
  "gid": 1,
  "rev": 0,
  "msg": "test",
  "app_proto": "smb",
  "requirements": [
    "flow",
    "dcerpc"
  ],
  "flags": [
    "src_any",
    "dst_any",
    "sp_any",
    "dp_any",
    "applayer",
    "toserver",
    "toclient",
    "prefilter"
  ],
  "pkt_engines": [],
  "frame_engines": [],
  "engines": [
    {
      "name": "dce_stub_data",
      "direction": "toclient",
      "is_mpm": true,
      "app_proto": "smb",
      "progress": 0,
      "matches": [
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": true,
            "no_double_inspect": false,
            "within": 32,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 6,
            "within": 12,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|03 00 00 00|",
            "length": 4,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 10,
            "within": 4,
            "fast_pattern": false
          }
        }
      ]
    },
    {
      "name": "dce_stub_data",
      "direction": "toserver",
      "is_mpm": true,
      "app_proto": "smb",
      "progress": 0,
      "matches": [
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": true,
            "no_double_inspect": false,
            "within": 32,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 6,
            "within": 12,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|03 00 00 00|",
            "length": 4,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 10,
            "within": 4,
            "fast_pattern": false
          }
        }
      ]
    },
    {
      "name": "dce_generic",
      "direction": "toserver",
      "is_mpm": false,
      "app_proto": "smb",
      "progress": 0,
      "matches": [
        {
          "name": "dcerpc.iface"
        },
        {
          "name": "dcerpc.opnum"
        }
      ]
    },
    {
      "name": "dce_generic",
      "direction": "toclient",
      "is_mpm": false,
      "app_proto": "smb",
      "progress": 0,
      "matches": [
        {
          "name": "dcerpc.iface"
        },
        {
          "name": "dcerpc.opnum"
        }
      ]
    }
  ],
  "lists": {},
  "mpm": {
    "buffer": "dce_stub_data",
    "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
    "length": 12,
    "nocase": false,
    "negated": false,
    "starts_with": false,
    "ends_with": false,
    "is_mpm": true,
    "no_double_inspect": false,
    "within": 32,
    "fast_pattern": false
  }
}
```

Reply:
same rule with `alert dcerpc` sets up for both `ALPROTO_DCERPC` and `ALPROTO_SMB`:
```json
{
  "raw": "alert dcerpc any any -> any any (msg:\"test\"; dcerpc.iface:367abb81-9844-35f1-ad32-98f038001003; dcerpc.opnum:12; dcerpc.stub_data; content:\"|15 00 00 00 00 00 00 00 15 00 00 00|\"; within:32; pcre:\"/^(?:[A-Z]\\x00){20}\\x00\\x00/R\"; content:\"|15 00 00 00 00 00 00 00 15 00 00 00|\"; distance:6; within:12; pcre:\"/^(?:[A-Z]\\x00){20}\\x00\\x00/R\"; content:\"|03 00 00 00|\"; distance:10; within:4; sid:6;)",
  "id": 6,
  "gid": 1,
  "rev": 0,
  "msg": "test",
  "app_proto": "dcerpc",
  "requirements": [
    "flow",
    "dcerpc"
  ],
  "flags": [
    "src_any",
    "dst_any",
    "sp_any",
    "dp_any",
    "applayer",
    "toserver",
    "toclient",
    "prefilter"
  ],
  "pkt_engines": [],
  "frame_engines": [],
  "engines": [
    {
      "name": "dce_stub_data",
      "direction": "toclient",
      "is_mpm": true,
      "app_proto": "dcerpc",
      "progress": 0,
      "matches": [
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": true,
            "no_double_inspect": false,
            "within": 32,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 6,
            "within": 12,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|03 00 00 00|",
            "length": 4,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 10,
            "within": 4,
            "fast_pattern": false
          }
        }
      ]
    },
    {
      "name": "dce_stub_data",
      "direction": "toserver",
      "is_mpm": true,
      "app_proto": "dcerpc",
      "progress": 0,
      "matches": [
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": true,
            "no_double_inspect": false,
            "within": 32,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 6,
            "within": 12,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|03 00 00 00|",
            "length": 4,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 10,
            "within": 4,
            "fast_pattern": false
          }
        }
      ]
    },
    {
      "name": "dce_stub_data",
      "direction": "toclient",
      "is_mpm": true,
      "app_proto": "smb",
      "progress": 0,
      "matches": [
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": true,
            "no_double_inspect": false,
            "within": 32,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 6,
            "within": 12,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|03 00 00 00|",
            "length": 4,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 10,
            "within": 4,
            "fast_pattern": false
          }
        }
      ]
    },
    {
      "name": "dce_stub_data",
      "direction": "toserver",
      "is_mpm": true,
      "app_proto": "smb",
      "progress": 0,
      "matches": [
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": true,
            "no_double_inspect": false,
            "within": 32,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
            "length": 12,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 6,
            "within": 12,
            "fast_pattern": false
          }
        },
        {
          "name": "pcre"
        },
        {
          "name": "content",
          "content": {
            "pattern": "|03 00 00 00|",
            "length": 4,
            "nocase": false,
            "negated": false,
            "starts_with": false,
            "ends_with": false,
            "is_mpm": false,
            "no_double_inspect": false,
            "distance": 10,
            "within": 4,
            "fast_pattern": false
          }
        }
      ]
    },
    {
      "name": "dce_generic",
      "direction": "toserver",
      "is_mpm": false,
      "app_proto": "dcerpc",
      "progress": 0,
      "matches": [
        {
          "name": "dcerpc.iface"
        },
        {
          "name": "dcerpc.opnum"
        }
      ]
    },
    {
      "name": "dce_generic",
      "direction": "toserver",
      "is_mpm": false,
      "app_proto": "smb",
      "progress": 0,
      "matches": [
        {
          "name": "dcerpc.iface"
        },
        {
          "name": "dcerpc.opnum"
        }
      ]
    },
    {
      "name": "dce_generic",
      "direction": "toclient",
      "is_mpm": false,
      "app_proto": "dcerpc",
      "progress": 0,
      "matches": [
        {
          "name": "dcerpc.iface"
        },
        {
          "name": "dcerpc.opnum"
        }
      ]
    },
    {
      "name": "dce_generic",
      "direction": "toclient",
      "is_mpm": false,
      "app_proto": "smb",
      "progress": 0,
      "matches": [
        {
          "name": "dcerpc.iface"
        },
        {
          "name": "dcerpc.opnum"
        }
      ]
    }
  ],
  "lists": {},
  "mpm": {
    "buffer": "dce_stub_data",
    "pattern": "|15 00 00 00 00 00 00 00 15 00 00 00|",
    "length": 12,
    "nocase": false,
    "negated": false,
    "starts_with": false,
    "ends_with": false,
    "is_mpm": true,
    "no_double_inspect": false,
    "within": 32,
    "fast_pattern": false
  }
}
```

Reply:
Yes, `alert smb` limits to `smb`. For `alert dcerpc` we have the special case where it applies both to "raw" dcerpc and dcerpc over smb.